### ISSUE TITLE
QuantSim accuracy is close to FP32, but inference result is incorrect when running with Genie (QNN 2.30, AIMET 2.0.0)

### ISSUE THREAD

**Issue by yuzhou164** [quic/aimet#4057]
QuantSim accuracy is close to FP32, but inference result is incorrect when running with Genie
**Environment**
QNN version: 2.30
AIMET version: 2.0.0
Runtime: Genie
Model type: LLM with KV-cache
Quantization method: AIMET QuantSim

**Accuracy Comparison**
After quantization, I evaluated the model accuracy and obtained the following results:
{
  "original": 3.7645184993743896,
  "prepared_kvcache": 3.7645175457000732,
  "QuantSim": 3.8976714611053467
}
Observations:
- The prepared KV-cache model matches the FP32 model very closely.
- The QuantSim model shows only a small accuracy gap compared to FP32.
Based on these results, the quantized model appears to be correct at the QuantSim level.

**Issue Description**
Although the QuantSim accuracy looks reasonable and close to the FP32 baseline, the inference behavior becomes abnormal when the quantized model is exported and executed using Genie.
This issue only appears at the Genie runtime stage and is not observed during QuantSim evaluation.
The abnormal behavior includes:
- Output quality is incorrect or inconsistent with expectations
- Results differ significantly from both FP32 and QuantSim outputs

**Expected Behavior**

My expectation is that: 
- If the QuantSim results are close to FP32,
- Then running the same quantized model with Genie should also produce correct or at  least comparable results.
I am not sure whether this assumption is valid.

**Questions**

1. Is it generally expected that good QuantSim results imply correct behavior when running the model with Genie?

2. What are common root causes for a model behaving correctly in QuantSim but incorrectly at Genie runtime, for example:
- Operator implementation or numerical differences
- Quantization encoding or configuration issues
- Export or graph transformation differences
- KV-cache-128 handling, layout, or update logic mismatch

3. What are the recommended steps or debugging tools to systematically locate and diagnose this kind of issue?

Any guidance or best practices for debugging QuantSim vs. Genie mismatches would be greatly appreciated.

**Comment by quic-ashvkuma** [comment]
Generally, those expectations should be valid. To answer your questions:

1. Good quantsim results imply that the the model is resilient to quantization. Since Genie is a full runtime, there are other issues that can arise not just related to quantization, which would therefore not be simulated in quantsim
2. My suspicion is that this could be related to how certain operators are represented in the exported graph. This is tough to debug further without more information though.

Would you be able to share more information about the model and script that you are running?

**Comment by yuzhou164** [comment]
I am using the AIMET tutorial provided by Qualcomm — tutorial_for_Qwen2.5-1.5B.
My environment versions are: aimet: 2.2.0 QNN: 2.40

**Step 1 Results**
When following Step 1 of the tutorial, I observe an issue with A16W4 quantization. The evaluation metrics are:
{
  "original": 3.7645184993743896,
  "prepared_kvcache": 3.7645175457000732,
  "QuantSim": 4.45306396484375
}
I then validated A16W8, and the results are:
{
  "original": 3.7645184993743896,
  "prepared_kvcache": 3.7645175457000732,
  "QuantSim": 3.896446943283081
}

**Step 2 – Step 3**
Step 2 and Step 3 were executed strictly according to the tutorial without any modifications.
However, after deployment and validation on Snapdragon 9100 using genie-net-run, the final generation quality is incorrect / significantly degraded for A16W4 weight Parameters. The result is correct for A16W8. 

**Additional Investigation**
I tried modifying the exception file by adding the following module patterns
"module_name": "\\w*rms_norm__model_layers_\\d+_.*" 
"module_name": "rms_norm__model_norm"。 
The reason is that I found the original pattern:\\w*rms_norm_\\d+ to be problematic in practice.

Test vector layes also have been modified 
 test_vector_layers = [
        "model_layers_\\d+_input_layernorm_Pow",
        "model_layers_\\d+_input_layernorm_Cast",
        "lm_head_conv_Conv",
        "lm_head_MatMul",
        "model.layers\\d+.input_layernorm.cast",
        "lm_head_conv",
        "lm_head",
        "model_embed_tokens_Gather",
        "rms_norm__model_layers_\\d+_post_attention_layernorm",
        "rms_norm__model_layers_\\d+_input_layernorm"
        "rms_norm__model_norm.weight",
        "model_embed_tokens_Gather",
        "model_layers_\\d+_Add_1",
        "rms_norm_\\d+"
    ]

 After applying these changes, the related configuration files accordingly (details below).

[exceptions.json](https://github.com/user-attachments/files/24563703/exceptions.json)

Running Genie common below:  ./qnn_340_genie/genie-t2t-run -c qwen2.5-1.5B/qwen2.5-1.5B-htp.json  -p "<|im_start|>system\nYou are a helpful AI Assistant<|im_end|><|im_start|>What is France's capital?\n<|im_end|>\n<|im_start|>assistant\n"

<img width="2096" height="377" alt="Image" src="https://github.com/user-attachments/assets/ab153f33-da4f-4380-86bc-7afe5df0381c" />.

**Comment by quic-ashvkuma** [comment]
Did you modify the notebook to use a different model than Qwen 2.5 1.5B? The PPL numbers you have shared are different than what I would expect for that model. If you are running on a different model, then there are several reasons why you could be seeing incorrect output: slightly different operators in the graph, different special tokens, etc.

**Comment by yuzhou164** [comment]
I am working on Qwen2.5-1.5B PTQ using AIMET (v2.2.0) + QNN 2.40, following the official AIMET tutorial.
Initially, I observed significant PPL drop when quantizing Qwen2.5-1.5B without applying SpinQuant-style outlier removal / rotation, so I integrated SpinQuant rotation before AIMET PTQ.

The infer coder is 
```
# coding=utf-8
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import datetime
from logging import Logger

import torch
import torch.distributed as dist
from transformers import Qwen2TokenizerFast
import transformers
from eval_utils.main import ptq_model
from eval_utils.modeling_qwen2 import Qwen2ForCausalLM
from utils import data_utils, eval_utils, utils
from utils.process_args import process_args_ptq

log: Logger = utils.get_logger("spinquant")

from eval_utils import gptq_utils, rotation_utils
from utils import data_utils, fuse_norm_utils, hadamard_utils, quant_utils, utils


def rotation_model(args, model, model_args=None):
    transformers.set_seed(args.seed)
    model.eval()
    # Rotate the weights
    if args.rotate:
        print(f"-----------------------------rotate----------------------------------------------------")
        fuse_norm_utils.fuse_layer_norms(model)
        rotation_utils.rotate_model(model, args)
        utils.cleanup_memory(verbos=True)

        quant_utils.add_actquant(model)  # Add Activation Wrapper to the model
        qlayers = quant_utils.find_qlayers(model)
        for name in qlayers:
            if "down_proj" in name:
                had_K, K = hadamard_utils.get_hadK(model.config.intermediate_size)
                qlayers[name].online_full_had = True
                qlayers[name].had_K = had_K
                qlayers[name].K = K
                qlayers[name].fp32_had = args.fp32_had
    return model

def infer(model,tokenizer, training_args):
    model.config.use_cache = False
    with torch.no_grad():
        prompt = "Human:请写下你对过去一年的总结，并提出自己在未来一年的目标和计划。\n\nAssistant:"
        inputs = tokenizer(prompt, padding=True, truncation=True, return_tensors="pt").to("cuda")
        model.to("cuda")
        generate_ids_inf = model.generate(inputs.input_ids, do_sample=False, max_length=training_args.model_max_length)
        print("="*100)
        print(tokenizer.batch_decode(generate_ids_inf, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0])
        print("="*100)
def train() -> None:
    dist.init_process_group(backend="nccl", timeout=datetime.timedelta(hours=8))
    model_args, training_args, ptq_args = process_args_ptq()
    local_rank = utils.get_local_rank()

    log.info("the rank is {}".format(local_rank))
    torch.distributed.barrier()

    config = transformers.AutoConfig.from_pretrained(
        model_args.input_model, token=model_args.access_token
    )
    # Llama v3.2 specific: Spinquant is not compatiable with tie_word_embeddings, clone lm_head from embed_tokens
    process_word_embeddings = False
    if config.tie_word_embeddings:
        config.tie_word_embeddings = False
        process_word_embeddings = True
    dtype = torch.bfloat16 if training_args.bf16 else torch.float16
    model = Qwen2ForCausalLM.from_pretrained(
        pretrained_model_name_or_path=model_args.input_model,
        config=config,
        torch_dtype=dtype,
        token=model_args.access_token,
    )
    if process_word_embeddings:
        model.lm_head.weight.data = model.model.embed_tokens.weight.data.clone()
    model.cuda()
    model.eval()
    # model = ptq_model(ptq_args, model, model_args)
    if ptq_args.rotate:
        model = rotation_model(ptq_args, model, model_args)
    
    for param in model.parameters(): 
        param.data = param.data.contiguous()

    model.seqlen = training_args.model_max_length
    if local_rank == 0:
        log.info("Model PTQ completed {}".format(model))
        log.info("Start to load tokenizer...")

    tokenizer = Qwen2TokenizerFast.from_pretrained(
        pretrained_model_name_or_path=model_args.input_model,
        cache_dir=training_args.cache_dir,
        model_max_length=training_args.model_max_length,
        padding_side="right",
        use_fast=True,
        add_eos_token=False,
        add_bos_token=False,
        token=model_args.access_token,
    )
    log.info("Complete tokenizer loading...")
    infer(model,tokenizer, training_args)
    
    config.tie_word_embeddings = True
    if ptq_args.save_no_had_model_path is not None:
        model.save_pretrained(ptq_args.save_no_had_model_path,safe_serialization=False)
        tokenizer.save_pretrained(ptq_args.save_no_had_model_path)
    del model
    model = Qwen2ForCausalLM.from_pretrained(
        pretrained_model_name_or_path=model_args.save_no_had_model_path,
        config=config,
        torch_dtype=dtype,
        token=model_args.access_token,
    )
    model.lm_head.weight.data = model.model.embed_tokens.weight.data.clone()
    infer(model,tokenizer, training_args)
    dist.barrier()


if __name__ == "__main__":
    train()
```
**The detailed outputs and logs are provided in the uploaded below**

[output.log](https://github.com/user-attachments/files/24633464/output.log)


**Observation 1: SpinQuant rotation with online Hadamard works correctly**
When I use the following SpinQuant-style rotation (with add_actquant + online Hadamard on down_proj), the model inference result is correct, and PPL improves significantly:

```
def rotation_model(args, model, model_args=None):
    transformers.set_seed(args.seed)
    model.eval()
    # Rotate the weights
    if args.rotate:
        print(f"-----------------------------rotate----------------------------------------------------")
        fuse_norm_utils.fuse_layer_norms(model)
        rotation_utils.rotate_model(model, args)
        utils.cleanup_memory(verbos=True)
        quant_utils.add_actquant(model)  # Add Activation Wrapper to the model
        
       qlayers = quant_utils.find_qlayers(model)
        for name in qlayers:
            if "down_proj" in name:
                had_K, K = hadamard_utils.get_hadK(model.config.intermediate_size)
                qlayers[name].online_full_had = True
                qlayers[name].had_K = had_K
                qlayers[name].K = K
                qlayers[name].fp32_had = args.fp32_had
    return model
```

This version:
Removes activation outliers effectively，Works well for inference，Matches SpinQuant paper behavior。

**Observation 2: Rotation without Hadamard causes incorrect results**
If I remove the following part:

```
quant_utils.add_actquant(...)
down_proj online Hadamard
```


and only keep:

```
def rotation_model(args, model, model_args=None):
    transformers.set_seed(args.seed)
    model.eval()

    if args.rotate:
        fuse_norm_utils.fuse_layer_norms(model)
        rotation_utils.rotate_model(model, args)
        utils.cleanup_memory(verbos=True)
```

Then:
- Inference results become incorrect / unstable
- PPL degrades badly
- Model cannot be reliably used as AIMET PTQ input

**Observation 3: rotate_model_no_had is referenced but not available**
In some SpinQuant-related code snippets, I found a pattern like:
`rotation_utils.rotate_model_no_had(model, args)`

Example:

```
fuse_norm_utils.fuse_layer_norms(model)
rotation_utils.rotate_model_no_had(model, args)

```

However:
- rotate_model_no_had is not available in the Aimet tutorial.
- It is unclear how this function differs from rotate_model.
- It is unclear whether this function performs a pure offline weight rotation suitable for deployment.

**Core Question**
My goal is:
To convert a SpinQuant-processed model into a static model that can be used as input to the AIMET PTQ tutorial (without online Hadamard / runtime activation hooks).

Specifically, I would like to ask:
1. Is the large PPL drop in AIMET PTQ expected if SpinQuant-style outlier removal is not applied for Qwen2.5-1.5B?
2. What is the correct way to export / save a SpinQuant-rotated model so that it can be consumed by AIMET PTQ?
3. Is there an official or recommended way to apply SpinQuant rotation offline only (no add_actquant, no online Hadamard)?
4. Does rotate_model_no_had exist internally, and if so, can it be shared or explained?

Thanks.
